File: python/red_black_tree.py
class Node:
  def __init__(self, data):
    self.data = data
    self.left = None
    self.right = None
    self.color = True 


class RedBlackTree:
  def __init__(self):
    self.root = None

  def insert(self, data):
    self.root = self._insert(self.root, data)
    self.root.color = False # root node is always black

  def insert(self, node, data):
    if node is None:
      return Node(data)

    if data < node.data:
      node.left = self._insert(node.left, data)
    elif data > node.data:
      node.right = self._insert(node.right, data)

    return self.balance(node)

  def balance(self, node):
    # left left case
    if self.isRed(node.left) and self.isRed(node.left.left):
      return self.rightRotate(node)

    # right right case
    if self.isRed(node.right) and self.isRed(node.right.right):
      return self.leftRotate(node)

    # left right case
    if self.isRed(node.left) and self.isRed(node.right):
      self.flipColors(node)

    return node

  def leftRotate(self, x):
    y = x.right
    x.right = y.left
    y.left = x
    y.color = x.color
    x.color = True
    return y


  def rightRotate(self, x):
    y = x.left
    x.left = y.right
    y.right = x
    y.color = x.color
    x.color = True
    return y


  def flipColors(self, node):
    node.color = not node.color
    node.left.color = not node.left.color
    node.right.color = not node.right.color 

  def isRed(self, node):
    if node is None:
      return False
    return node.color 

File: java/RedBlackTree.java
public class RedBlackTree<T extends Comparable<T>>
  {
    private Node<T> root;
    private static class Node<T> {
      T data;
      Node<T> left, right;
      boolean color;

      Node(T data) {
        this.data = data;
        this.color = true;
      }
    }
    public void insert(T data)
    {
      root = insert(root, data);
      root.color = false;
    }
    private Node<T> insert(Node<T> node, T data) {
      if(node == null) {
        return new Node<>(data);
      }
      if(data.compareTo(node.data) < 0) {
        node.left = insert(node.left, data);
      }
      else if(data.compareTo(node.data) > 0) {
        node.right = insert(node.right, data);
      }
      else {
      }
      return balance(node);
    }
    private Node<T> balance(Node<T> node)
    {
      if(isRed(node.right) && !isRed(node.left)) {
        node = rotateLeft(node);
      }
      if(isRed(node.left) && isRed(node.left.left)) {
        node = rotateRight(node);
      }
      if(isRed(node.left) && isRed(node.right)) {
        flipColors(node);
      }
      return node;
    }
    private Node<T> rotateLeft(Node<T> x) {
      Node<T> y = x.right;
      x.right = y.left;
      y.left = x;
      y.color = x.color;
      x.color = true;
      return y;
    }
    private Node<T> rotateRight(Node<T> x) {
      Node<T> y = x.left;
      x.left = y.right;
      y.right = x;
      y.color = x.color;
      x.color = true;
      return y;
    }
    private void flipColors(Node<T> node)
    {
      node.color = !node.color;
      node.left.color = !node.left.color;
      node.right.color = !node.right.color;
    }
    private boolean isRed(Node<T> x) {
      if(x == null) {
        return false;
      }
      return x.color;
    }
  }
